src/render.ts

```diff
@@ -1,63 +1,213 @@
+  });
+}
+
+function updateDOMProperties(
+  dom: Node,
+  currentProperties: JSXProps,
+  updateProperties: JSXProps
+) {
+  if (!(dom instanceof HTMLElement)) return;
+
+  // 현재 속성에서 업데이트된 속성과 비교하여 이벤트 리스너를 제거
+  for (const [key, value] of Object.entries(currentProperties)) {
+    if (key === "children" || key === "key") continue;
+    if (!(key in updateProperties)) {
+      // 이벤트 리스너 삭제
+      if (key.startsWith("on")) {
+        const eventName = key.slice(2).toLowerCase();
+        const listener = getEventListeners(dom)?.get(eventName);
+        if (listener) {
+          dom.removeEventListener(eventName, listener);
+          getEventListeners(dom)?.delete(eventName); // 기존 리스너 제거
         }
+      } else {
+        // 다른 속성 삭제
+        dom.removeAttribute(key);
+      }
+    }
+  }
+
+  // 업데이트된 속성에 대해서 새 이벤트 리스너를 추가하거나 속성을 설정
+  for (const [key, value] of Object.entries(updateProperties)) {
+    if (key === "children" || key === "key") continue;
+    if (!(dom instanceof HTMLElement)) return;
+
+    const oldValue = currentProperties[key];
+    if (oldValue !== value) {
+      if (key === "className") {
+        dom.setAttribute("class", value.toString());
+      } else if (key.startsWith("on") && typeof value === "function") {
+        const eventName = key.slice(2).toLowerCase();
+        const listener = (event: Event) => {
+          const syntheticEvent = createSyntheticEvent(event);
+          (value as (e: SyntheticEventProps) => void)(syntheticEvent);
+        };
+
+        // 기존 리스너 제거
+        cleanupEventListener(dom);
+
+        // 새 리스너 추가
+        dom.addEventListener(eventName, listener);
+        addEventListenerToMap(dom, eventName, listener);
+      } else {
+        dom.setAttribute(key, value.toString());
+      }
+    }
+  }
+}
+
+
+function updateChildren(dom, currentChildren, updateChildren) {
+  const currentChildArray = Array.isArray(currentChildren)
+    ? currentChildren
+    : [currentChildren];
+  const updateChildArray = Array.isArray(updateChildren)
+    ? updateChildren
+    : [updateChildren];
+
+  const currentNodes = Array.from(dom.childNodes);
+  currentChildArray.forEach((child, index) => {
+    if (!updateChildArray[index]) {
+      dom.removeChild(currentNodes[index]);
+    } else {
+      const newDom = reconcile(dom, child, updateChildArray[index]);
+      if (newDom !== currentNodes[index]) {
+        dom.replaceChild(newDom, currentNodes[index]);
+      }
     }
+  });
+
+  for (let i = currentChildArray.length; i < updateChildArray.length; i++) {
+    dom.appendChild(createRealDOM(updateChildArray[i]));
+  }
+}
+
+function createRealDOM(virtualDOM) {
+  if (typeof virtualDOM === "string" || typeof virtualDOM === "number") {
+    return document.createTextNode(virtualDOM.toString());
   }
 
+  const { type, props } = virtualDOM;
+  const dom = document.createElement(type);
+  updateDOMProperties(dom, {}, props);
+  updateChildren(dom, [], props.children || []);
```

Comment:
P555;

위에 언급한 내용의 예시로 이런 경우 저는 의미있는 변수로 어떤 값이 들어가는지 명시하는 편이에요.

```suggestion
  const children = props.children ?? [];
  updateChildren(dom, [], children);
```